extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/caffeine/CaffeineCache.java

```diff
@@ -0,0 +1,158 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+
+import com.github.benmanes.caffeine.cache.Caffeine;
+
+import io.quarkus.cache.runtime.Cache;
+
+public class CaffeineCache implements Cache {
+
+    private com.github.benmanes.caffeine.cache.Cache<Object, Object> cache;
+
+    private String name;
+
+    private Integer initialCapacity;
+
+    private Long maximumSize;
+
+    private Duration expireAfterWrite;
+
+    private Duration expireAfterAccess;
+
+    public CaffeineCache(CaffeineCacheInfo cacheInfo) {
+        this.name = cacheInfo.name;
+        Caffeine<Object, Object> builder = Caffeine.newBuilder();
+        // TODO: The following line is a workaround for a GraalVM issue: https://github.com/oracle/graal/issues/1524
+        // It should be removed as soon as a Quarkus release depends on GraalVM 19.3.0 or greater.
+        builder.executor(Runnable::run);
+        if (cacheInfo.initialCapacity != null) {
+            this.initialCapacity = cacheInfo.initialCapacity;
+            builder.initialCapacity(cacheInfo.initialCapacity);
+        }
+        if (cacheInfo.maximumSize != null) {
+            this.maximumSize = cacheInfo.maximumSize;
+            builder.maximumSize(cacheInfo.maximumSize);
+        }
+        if (cacheInfo.expireAfterWrite != null) {
+            this.expireAfterWrite = cacheInfo.expireAfterWrite;
+            builder.expireAfterWrite(cacheInfo.expireAfterWrite);
+        }
+        if (cacheInfo.expireAfterAccess != null) {
+            this.expireAfterAccess = cacheInfo.expireAfterAccess;
+            builder.expireAfterAccess(cacheInfo.expireAfterAccess);
+        }
+        cache = builder.build();
+    }
+
+    @Override
+    public Object get(Object key, Callable<Object> valueLoader, long lockTimeout) throws Exception {
+        if (lockTimeout <= 0) {
+            return fromCacheValue(cache.get(key, new MappingFunction(valueLoader)));
+        }
+
+        // The lock timeout logic starts here.
+
+        /*
+         * If the current key is not already associated with a value in the Caffeine cache, there's no way to know if the
+         * current thread or another one is responsible for computing the missing value. While we want to wait at most for the
+         * lock timeout delay if the value is being computed by another thread, we also want to ignore the timeout if the value
+         * is being computed from the current thread. The following variable will be used to make sure we don't interrupt a
+         * current-thread computation.
+         */
+        AtomicBoolean isCurrentThreadComputation = new AtomicBoolean();
+
+        CompletableFuture<Object> future = CompletableFuture.supplyAsync(() -> {
+            return fromCacheValue(cache.get(key, new MappingFunction(valueLoader, isCurrentThreadComputation)));
+        });
```

Comment:
Why not simply use an `AsyncCache`? Then you can write this as,

```java
boolean[] isCurrentThreadComputation = { false };
var future = cache.get(key, (k, executor) -> {
  isCurrentThreadComputation[0] = true;
  return CompletableFuture.supplyAsync(() -> toCacheValue(valueLoader.call()), executor);
});
return isCurrentThreadComputation[0] ? future.get() : future.get(lockTimeout, TimeUnit.MILLISECONDS);
```

Reply:
I thought about it and replaced `Cache` with `AsyncCache` at some point, but then I decided to move back to `Cache` to avoid paying the cost of `AsyncCache` for all cache loading operations while the `CompletableFuture` part was only needed when the `lockTimeout` parameter is present in `@CacheResult` (it is disabled by default).

I also went for a double-get version of the code because I was afraid `isCurrentThreadComputation[0]` would be evaluated before the `CompletableFuture` had a chance to start its computation.

Reply:
That's a fair concern, but requires being extra careful about its usage. In your approach every cache access requires a context switch so that a different thread performs the operation and notifies the caller. This adds pressure to both the `ForkJoinPool.commonPool()` and the GC, as allocations and available threads are required.

You can reduce much of this for the happy path of a read by using an optimistic lookup (`getIfPresent`) and fallback to the future-based code.

For a cache load, you are causing the future not performing the work to be blocked on the cache's internal locking. This can exhaust the thread pool, which may spin up extra threads (see `ManagedBlocker`) to compensate.

The simplest and least error prone, imho, would be to use `Cache` if `lockTimeout=0` and `AsyncCache` otherwise. The extra memory overhead would only be for those who enable timeouts or (if you wanted to support) returning a future. I think those two variants would strike a better balance, keep the code simple, and avoid nasty surprises.

Reply:
Thanks for your precious advice @ben-manes, the thread pool exhaustion needed to be fixed indeed. I've pushed some changes, the new code is now relying on an `AsyncCache` which is used in a synchronized and asynchronized way depending on the context. Does it look better that way?

Quarkus still depends on Caffeine 2.6.x for now, I need to figure out how to move to 2.7 or 2.8 but there are some dependencies conflicts that need to be solved first. That's why I used a workaround while initializing the `AsyncLoadingCache`. I'll fix this before thinking about changing the double-get logic (I'm still not sure about it).

Reply:
Thanks, that does look better.

I would propose a few minor changes, but probably defer since you're tackling so much in this PR.
1. The sample I showed set the flag within the mapping function, before the future was returned. Internally the cache holds a `ConcurrentHashMap<K, CompletableFuture<V>>`, so this is an atomic computation. That ensures that the flag is set, regardless of when the future is executed. In your case the race condition will certainly occur if read immediately or the pool is backed up, but is a rather harmless.
2. An optimization would be to have two `get` methods variants based on if the cache is constructed as synchronous or async. Then you can avoid your per-entry overhead in the common case, which very little code complexity.

Reply:
I made a few changes inspired from [1], thanks for the suggestion!

Reply:
nice. Another to-do list item for a future PR would be,
3. Avoid delegating to the thread pool by supplying an incomplete future. The computing thread would then fulfil it, rather than block.